### PR TITLE
CUDA, TF, Pytorch version upgrade

### PR DESCRIPTION
**Pull request by ssheorey**
- CUDA 10.1 -> 11.0
- Torch 1.6.0 -> 1.7.1
- TF 2.3.1 -> 2.4.1
- GPU CI upgrades: NVIDIA driver 440 -> 460

NPP cmake library checks

Corresponding Open3D-ML PR: https://github.com/intel-isl/Open3D-ML/pull/224

---
This change is [Reviewable](https://reviewable.io/reviews/intel-isl/open3d/3049)

### CODE REVIEW

**Comment by ssheorey** [comment]
Python test failure on RTX3070: 
`python/test/ml_ops/test_voxelize.py` crashes:
```

> 2021-02-16 13:11:56.347720: E tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:725] failed to record completion event; therefore, failed to create inter-stream dependency
> 2021-02-16 13:11:56.347739: E tensorflow/stream_executor/cuda/cuda_driver.cc:1071] failed to enqueue async memcpy from host to device: CUDA_ERROR_ILLEGAL_ADDRESS: an illegal memory access was encountered; GPU dst: 0x7f6f9aa00600; host src: 0x55b3bc2e0b40; size: 28656=0x6ff0
> 2021-02-16 13:11:56.347745: E tensorflow/stream_executor/stream.cc:334] Error recording event in stream: Error recording CUDA event: CUDA_ERROR_ILLEGAL_ADDRESS: an illegal memory access was encountered; not marking stream as bad, as the Event object may be at fault. Monitor for further errors.
> 2021-02-16 13:11:56.347753: E tensorflow/stream_executor/cuda/cuda_event.cc:29] Error polling for event status: failed to query event: CUDA_ERROR_ILLEGAL_ADDRESS: an illegal memory access was encountered
> 2021-02-16 13:11:56.347761: F tensorflow/core/common_runtime/gpu/gpu_event_mgr.cc:220] Unexpected Event status: 1
> Fatal Python error: Aborted

Current thread 0x00007f71d2229280 (most recent call first):
  File "<string>", line 2661 in open3d_voxelize
  File "/home/ssheorey/Documents/Open3D/Code/Open3D/build/lib/python_package/open3d/ml/tf/python/ops/ops.py", line 1595 in voxelize
  File "/home/ssheorey/Documents/Open3D/Code/Open3D/python/test/ml_ops/mltest.py", line 78 in run_op
  File "/home/ssheorey/Documents/Open3D/Code/Open3D/python/test/ml_ops/test_voxelize.py", line 162 in test_voxelize_random
```

Open3D-ML test failure in `test_models.py`

```
> 2021-02-16 13:12:03.662251: I tensorflow/stream_executor/stream.cc:1404] [stream=0x55ba6d4466c0,impl=0x55ba6d443930] did not wait for [stream=0x55bb1aa0ea40,impl=0x55ba6d443780]
> 2021-02-16 13:12:03.662420: F tensorflow/core/common_runtime/gpu/gpu_util.cc:340] CPU->GPU Memcpy failed
> Fatal Python error: Aborted

Thread 0x00007f7d8311f280 (most recent call first):
  File "/home/ssheorey/.pyenv/versions/open3d/lib/python3.8/site-packages/tensorflow/python/ops/gen_math_ops.py", line 6064 in mul
  File "/home/ssheorey/.pyenv/versions/open3d/lib/python3.8/site-packages/tensorflow/python/ops/math_ops.py", line 518 in multiply
  File "/home/ssheorey/.pyenv/versions/open3d/lib/python3.8/site-packages/tensorflow/python/util/dispatch.py", line 201 in wrapper
  File "/home/ssheorey/.pyenv/versions/open3d/lib/python3.8/site-packages/tensorflow/python/ops/math_ops.py", line 1496 in _mul_dispatch
  File "/home/ssheorey/.pyenv/versions/open3d/lib/python3.8/site-packages/tensorflow/python/ops/math_ops.py", line 1164 in binary_op_wrapper
  File "/home/ssheorey/.pyenv/versions/open3d/lib/python3.8/site-packages/tensorflow/python/ops/random_ops.py", line 96 in random_normal
  File "/home/ssheorey/.pyenv/versions/open3d/lib/python3.8/site-packages/tensorflow/python/util/dispatch.py", line 201 in wrapper
  File "/home/ssheorey/.pyenv/versions/open3d/lib/python3.8/site-packages/tensorflow/python/ops/init_ops_v2.py", line 1072 in random_normal
  File "/home/ssheorey/.pyenv/versions/open3d/lib/python3.8/site-packages/tensorflow/python/ops/init_ops_v2.py", line 419 in __call__
  File "/home/ssheorey/.pyenv/versions/open3d/lib/python3.8/site-packages/tensorflow/python/keras/initializers/initializers_v2.py", line 304 in __call__
  File "/home/ssheorey/.pyenv/versions/open3d/lib/python3.8/site-packages/tensorflow/python/ops/resource_variable_ops.py", line 1712 in _init_from_args
  File "/home/ssheorey/.pyenv/versions/open3d/lib/python3.8/site-packages/tensorflow/python/ops/resource_variable_ops.py", line 1574 in __init__
  File "/home/ssheorey/.pyenv/versions/open3d/lib/python3.8/site-packages/tensorflow/python/ops/variables.py", line 264 in __call__
  File "/home/ssheorey/.pyenv/versions/open3d/lib/python3.8/site-packages/tensorflow/python/ops/variable_scope.py", line 2604 in default_variable_creator
  File "/home/ssheorey/.pyenv/versions/open3d/lib/python3.8/site-packages/tensorflow/python/ops/variables.py", line 199 in <lambda>
  File "/home/ssheorey/.pyenv/versions/open3d/lib/python3.8/site-packages/tensorflow/python/ops/variables.py", line 206 in _variable_v1_call
  File "/home/ssheorey/.pyenv/versions/open3d/lib/python3.8/site-packages/tensorflow/python/ops/variables.py", line 260 in __call__
  File "/home/ssheorey/.pyenv/versions/open3d/lib/python3.8/site-packages/tensorflow/python/keras/engine/base_layer_utils.py", line 130 in make_variable
  File "/home/ssheorey/.pyenv/versions/open3d/lib/python3.8/site-packages/tensorflow/python/training/tracking/base.py", line 805 in _add_variable_with_custom_getter
  File "/home/ssheorey/.pyenv/versions/open3d/lib/python3.8/site-packages/tensorflow/python/keras/engine/base_layer.py", line 623 in add_weight
  File "/home/ssheorey/Documents/Open3D/Code/Open3D-ML/ml3d/tf/models/network_blocks.py", line 174 in __init__
  File "/home/ssheorey/Documents/Open3D/Code/Open3D-ML/ml3d/tf/models/network_blocks.py", line 509 in __init__
  File "/home/ssheorey/Documents/Open3D/Code/Open3D-ML/ml3d/tf/models/network_blocks.py", line 92 in block_decider
  File "/home/ssheorey/Documents/Open3D/Code/Open3D-ML/ml3d/tf/models/kpconv.py", line 173 in __init__
  File "/home/ssheorey/Documents/Open3D/Code/Open3D-ML/tests/test_models.py", line 117 in test_kpconv_tf


```

**Comment by ssheorey** [comment]
```
> [ RUN      ] Hashmap/HashmapPermuteDevices.Find/1
> terminate called after throwing an instance of 'thrust::system::system_error'
>   what():  CUDA free failed: cudaErrorIllegalAddress: an illegal memory access was encountered
```

File: CMakeLists.txt [hunk 1]
```diff
@@ -654,6 +657,11 @@ if(BUILD_PYTHON_MODULE)
 endif()
 open3d_aligned_print("Build TensorFlow Ops" "${BUILD_TENSORFLOW_OPS}")
 open3d_aligned_print("Build PyTorch Ops" "${BUILD_PYTORCH_OPS}")
+if (BUILD_PYTORCH_OPS AND BUILD_CUDA_MODULE)
+    message(WARNING "PyTorch Ops with CUDA needs a custom PyTorch install from
+    https://github.com/intel-isl/open3d_downloads/releases/tag/torch1.7.1 The
+    official PyTorch v1.7.1 wheel will not work due to issue #52663")
```
Comment (yxlao) [inline CMakeLists.txt:663]: shall we provide a link to the issue directly?

File: util/ci_utils.sh [hunk 2]
```diff
@@ -51,13 +54,20 @@ OPEN3D_SOURCE_ROOT="$(cd "$(dirname "${BASH_SOURCE[0]}")"/.. >/dev/null 2>&1 &&
...
-    $SUDO apt-get install --yes --no-install-recommends \
-        "cuda-toolkit-${CUDA_VERSION[0]}" \
-        libcublas-dev
+    options="$(echo "$@" | tr ' ' '|')"
+
+    if [[ $UBUNTU_VERSION == "bionic" ]]; then
+        echo "Installing CUDA ${CUDA_VERSION[1]} with apt from bionic repos..."
+        $SUDO apt-key adv --fetch-keys https://developer.download.nvidia.com/compute/cuda/repos/ubuntu1804/x86_64/7fa2af80.pub
+        $SUDO apt-add-repository "deb https://developer.download.nvidia.com/compute/cuda/repos/ubuntu1804/x86_64 /"
+    else
```
Comment (yxlao) [inline util/ci_utils.sh:65]: we could consider:
```
if (18.04) {...}
else if (20.04) {...}
else {error}
```

File: util/ci_utils.sh [hunk 3]
```diff
@@ -75,41 +85,40 @@ install_cuda_toolkit() {
...
-        $SUDO tar --no-same-owner -xzf cudnn-10.1-linux-x64-v7.6.5.32.tgz -C /usr/local
-        rm cudnn-10.1-linux-x64-v7.6.5.32.tgz
-        $SUDO ldconfig
-        # We may revisit the repository approach in the future.
-        # echo "Installing cuDNN ${CUDNN_VERSION} with apt ..."
-        # $SUDO apt-add-repository "deb https://developer.download.nvidia.com/compute/machine-learning/repos/ubuntu1804/x86_64 /"
-        # $SUDO apt-get install --yes --no-install-recommends \
-        #     "libcudnn${CUDNN_MAJOR_VERSION}=$CUDNN_VERSION" \
-        #     "libcudnn${CUDNN_MAJOR_VERSION}-dev=$CUDNN_VERSION"
+        #$SUDO apt-get install --yes --no-install-recommends curl
```
Comment (yxlao) [inline util/ci_utils.sh:88]: nit: `# `  (add space after), if we want to keep the comments

**Review by yxlao — APPROVED** [review]

File: util/ci_utils.sh (hunk 2 above)
Comment (ssheorey) [reply util/ci_utils.sh]: Done

File: CMakeLists.txt (hunk 1 above)
Comment (ssheorey) [reply CMakeLists.txt]: @benjaminum updated to a much nicer warning. Also added that in the docs.

File: util/ci_utils.sh (hunk 3 above)
Comment (ssheorey) [reply util/ci_utils.sh]: Done

**Review by benjaminum — APPROVED** [review]
> Reviewed 2 of 8 files at r1, 6 of 10 files at r4, 4 of 4 files at r6.
> *[Reviewable](https://reviewable.io/reviews/intel-isl/open3d/3049)* status: all files reviewed, 2 unresolved discussions (waiting on @benjaminum, @stotko, and @yxlao)

**Review by stotko — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/intel-isl/open3d/3049)* status: all files reviewed, 4 unresolved discussions (waiting on @benjaminum, @ssheorey, and @yxlao)

___
*[docs/compilation.rst, line 25 at r6](https://reviewable.io/reviews/intel-isl/open3d/3049#-MZHMOM5BMZSg-GtB-QG:-MZHMOM68Exb7xeyv7ts:bvebugl) ([raw file](https://github.com/intel-isl/open3d/blob/0cf73758f7edd7f578b2199db7fc07dc465197af/docs/compilation.rst#L25)):*
> ```reStructuredText
>   * Windows: Download from: `CMake download page <https://cmake.org/download/>`_
> 
> * CUDA 11.0 (optional): Open3D supports GPU acceleration of an increasing number
> ```

This suggests that CUDA 11.0 is the *minimum* required version. Alternatively, it could also be interpreted as this is the *exact* required version. Maybe, we should mention a range of supported versions because the newest one is bound by the Pytorch and Tensorflow releases, right? So something like `CUDA 10.1-11.0` (what about `11.1`?) and then adding bullet points for Pytorch and Tensorflow after that with the hint that for GPU support the CUDA version has to match?
___
*[python/test/ml_ops/test_sparseconv.py, line 148 at r6](https://reviewable.io/reviews/intel-isl/open3d/3049#-MZHO2nH6hdibnOG0nSe:-MZHO2nH6hdibnOG0nSf:b9v7pn) ([raw file](https://github.com/intel-isl/open3d/blob/0cf73758f7edd7f578b2199db7fc07dc465197af/python/test/ml_ops/test_sparseconv.py#L148)):*
> ```Python
>     y_conv3d += bias
> 
>     np.testing.assert_allclose(y, y_conv3d, rtol=1e-3, atol=1e-5)
> ```

Are these still needed? I have not observed any failures on my local machine with CUDA 11 without these changes.

**Review by benjaminum — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/intel-isl/open3d/3049)* status: all files reviewed, 4 unresolved discussions (waiting on @benjaminum, @ssheorey, @stotko, and @yxlao)

___
*[python/test/ml_ops/test_sparseconv.py, line 148 at r6](https://reviewable.io/reviews/intel-isl/open3d/3049#-MZHO2nH6hdibnOG0nSe:-MZISois98_jW4ONAijf:b55084b) ([raw file](https://github.com/intel-isl/open3d/blob/0cf73758f7edd7f578b2199db7fc07dc465197af/python/test/ml_ops/test_sparseconv.py#L148)):*
<details><summary><i>Previously, stotko (Patrick Stotko) wrote…</i></summary><blockquote>

Are these still needed? I have not observed any failures on my local machine with CUDA 11 without these changes.
</blockquote></details>

The accuracy seems to depend on the driver and the GPU. Without the changes the test works for me on T4 in GCE and my workstation with 2080ti but fails on the cluster with a 2080ti (same driver as GCE).

**Review by benjaminum — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/intel-isl/open3d/3049)* status: all files reviewed, 4 unresolved discussions (waiting on @benjaminum, @stotko, and @yxlao)

___
*[docs/compilation.rst, line 25 at r6](https://reviewable.io/reviews/intel-isl/open3d/3049#-MZHMOM5BMZSg-GtB-QG:-MZIpWmU48Uusa_XqLKw:b-sfnfwz) ([raw file](https://github.com/intel-isl/open3d/blob/0cf73758f7edd7f578b2199db7fc07dc465197af/docs/compilation.rst#L25)):*
<details><summary><i>Previously, stotko (Patrick Stotko) wrote…</i></summary><blockquote>

This suggests that CUDA 11.0 is the *minimum* required version. Alternatively, it could also be interpreted as this is the *exact* required version. Maybe, we should mention a range of supported versions because the newest one is bound by the Pytorch and Tensorflow releases, right? So something like `CUDA 10.1-11.0` (what about `11.1`?) and then adding bullet points for Pytorch and Tensorflow after that with the hint that for GPU support the CUDA version has to match?
</blockquote></details>

Everything from CUDA 10 to CUDA 11.3 should work (as long as users install or compile tensorflow or pytorch with the same CUDA). I think our problem here is that we only test CUDA 11.0.
Would this be better
```
* CUDA (optional): Open3D supports GPU acceleration of an increasing number
  of operations through CUDA on Linux. We recommend to use CUDA 11.0 for the
  best compatibility with recent GPUs and optional external dependencies such
  as Tensorflow or PyTorch.
  Please see the `official documentation
  <https://docs.nvidia.com/cuda/cuda-installation-guide-linux/index.html>`_ to
  install the CUDA toolkit from Nvidia.
```
